### PR TITLE
add SSH Module

### PR DESCRIPTION
**Pull request by MambaWong**
add SSH Module.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @MambaWong ,
Thanks for your kind contribution. Would you like to fix the lint issue so that we could merge this PR?
The details can be found by clicking the item of lint check as shown below
image

**Comment by MambaWong** [comment]
@ZwwWayne need upgrade pre commit hooks described in #8963 ？

File: mmdet/models/necks/ssh.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+class SSHModule(BaseModule):
+    """This is an implementation of `SSH module` described in `SSH: Single
+    Stage Headless Face Detector <https://arxiv.org/pdf/1708.03979.pdf>`_.
+
+    Args:
+        in_channels (int): Number of input channels used at each scale.
+        out_channels (int): Number of output channels used at each scale.
+        conv_cfg (dict, optional): Config dict for convolution layer.
+            Default: None, which means using conv2d.
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:21]:
```suggestion
            Defaults to None.

File: mmdet/models/necks/ssh.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        norm_cfg (dict): Config dict for normalization layer.
+            Defaults to dict(type='BN').
+        init_cfg (dict or list[dict], optional): Initialization config dict.
+    """
+
+    def __init__(self,
+                 in_channels,
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:28]: Add type-hint here, you can follow other file to update arges here.

File: mmdet/models/necks/ssh.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    """`SSH Neck` used in `SSH: Single Stage Headless Face Detector.
+
+    <https://arxiv.org/pdf/1708.03979.pdf>`_.
+
+    Args:
+        num_scales (int): The number of scales / stages.
+        in_channels (List[int]): The number of input channels per scale.
+        out_channels (List[int]): The number of output channels  per scale.
+        conv_cfg (dict, optional): Config dict for convolution layer.
+            Default: None.
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:113]:
```suggestion
            Defaults to None.
```

File: mmdet/models/necks/ssh.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        norm_cfg (dict, optional): Dictionary to construct and config norm
+            layer. Default: dict(type='BN', requires_grad=True)
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:115]:
```suggestion
            layer. Defaults to dict(type='BN', requires_grad=True)
```

File: mmdet/models/necks/ssh.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        init_cfg (dict or list[dict], optional): Initialization config dict.
+            Default: None
+    """
+
+    def __init__(self,
+                 num_scales,
+                 in_channels: List[int],
+                 out_channels: List[int],
+                 conv_cfg: OptConfigType = None,
+                 norm_cfg: ConfigType = dict(type='BN'),
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:125]: dict(type='BN') or dict(type='BN', requires_grad=True). should unified.

File: mmdet/models/necks/ssh.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                 init_cfg: ConfigType = dict(
+                     type='Xavier', layer='Conv2d', distribution='uniform')):
+        super(SSH, self).__init__(init_cfg=init_cfg)
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:128]:
```suggestion
        super().__init__(init_cfg=init_cfg)
```

File: mmdet/models/necks/ssh.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        init_cfg (dict or list[dict], optional): Initialization config dict.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 conv_cfg=None,
+                 norm_cfg: ConfigType = dict(type='BN'),
+                 init_cfg=None):
+        super(SSHModule, self).__init__(init_cfg)
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:33]:
```suggestion
        super().__init__(init_cfg=init_cfg)
```

File: mmdet/models/necks/ssh.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+@MODELS.register_module()
+class RetinaSSH(BaseModule):
+    """`SSH Neck` used in `RetinaFace: Single-stage Dense Face Localisation in
+    the Wild <https://arxiv.org/pdf/1905.00641.pdf>`_.
+
+    Args:
+        in_channels (int): Number of input channels used at each scale.
+        out_channels (int): Number of output channels used at each scale.
+        conv_cfg (dict, optional): Config dict for convolution layer.
+            Default: None, which means using conv2d.
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:161]:
```suggestion
            Defaults to None.
```

File: mmdet/models/necks/ssh.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 conv_cfg=None,
+                 norm_cfg: ConfigType = dict(type='BN'),
+                 init_cfg: ConfigType = dict(
+                     type='Xavier', layer='Conv2d', distribution='uniform')):
+        super(RetinaSSH, self).__init__(init_cfg)
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:174]:
```suggestion
        super().__init__(init_cfg=init_cfg)
```

**Comment by BIGWangYuDong** [comment]
Maybe we can add some config demo in configs.

File: mmdet/models/necks/ssh.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                 norm_cfg: ConfigType = dict(type='BN'),
+                 init_cfg: OptMultiConfig = dict(
+                     type='Xavier', layer='Conv2d', distribution='uniform')):
+        super().__init__(init_cfg=init_cfg)
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        self.ssh_module = SSHModule(
+            in_channels, out_channels, conv_cfg=conv_cfg, norm_cfg=norm_cfg)
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:183]: in_channels=in_channels, out_channels=out_channels

File: mmdet/models/necks/ssh.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        assert (num_scales == len(in_channels) == len(out_channels))
+        self.num_scales = num_scales
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        for idx in range(self.num_scales):
+            in_c, out_c = self.in_channels[idx], self.out_channels[idx]
+            self.add_module(
+                f'ssh_module{idx}',
+                SSHModule(in_c, out_c, conv_cfg=conv_cfg, norm_cfg=norm_cfg))
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:140]: in_channel=in_c, out_channel=out_c

File: mmdet/models/necks/ssh.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+                 conv_cfg: OptConfigType = None,
+                 norm_cfg: ConfigType = dict(type='BN'),
+                 init_cfg: OptMultiConfig = None):
+        super().__init__(init_cfg=init_cfg)
+        assert out_channels % 4 == 0
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        self.conv3X3 = ConvModule(
```
Comment (hhaAndroid) [inline mmdet/models/necks/ssh.py:41]: self.conv3x3

File: mmdet/models/necks/ssh.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+import torch
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule
+from mmengine.model import BaseModule
+
+from mmdet.registry import MODELS
+from mmdet.utils import ConfigType, OptConfigType, OptMultiConfig
+
+
+class SSHModule(BaseModule):
```
Comment (hhaAndroid) [inline mmdet/models/necks/ssh.py:13]: SSHDetModule

File: mmdet/models/necks/ssh.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+            self.in_channels,
+            self.out_channels // 2,
+            3,
+            stride=1,
+            padding=1,
+            conv_cfg=conv_cfg,
+            norm_cfg=norm_cfg,
+            act_cfg=None)
+
+        self.conv5X5_1 = ConvModule(
```
Comment (hhaAndroid) [inline mmdet/models/necks/ssh.py:51]: Context Module

File: mmdet/models/necks/ssh.py [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+            self.out_channels // 4,
+            3,
+            stride=1,
+            padding=1,
+            conv_cfg=conv_cfg,
+            norm_cfg=norm_cfg,
+            act_cfg=None,
+        )
+
+    def forward(self, x):
```
Comment (hhaAndroid) [inline mmdet/models/necks/ssh.py:92]: It seems a little different from the paper?

Comment (MambaWong) [reply mmdet/models/necks/ssh.py:92]: Can you point out what you think is different?

File: mmdet/models/necks/ssh.py (hunk 13 above)
Comment (MambaWong) [reply mmdet/models/necks/ssh.py]: decompose the SSHModule into SSHContextModule and SSHDetModule

File: mmdet/models/necks/ssh.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import List, Tuple
```
Comment (hhaAndroid) [inline mmdet/models/necks/ssh.py:2]: Please add UT

File: mmdet/models/necks/ssh.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+            convolution layer. Defaults to None.
+        norm_cfg (:obj:`ConfigDict` or dict): Config dict for normalization
+            layer. Defaults to dict(type='BN').
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 conv_cfg: OptConfigType = None,
+                 norm_cfg: ConfigType = dict(type='BN')):
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/ssh.py:32]: should add init_cfg here, and set super().__init__(init_cfg=init_cfg)

**Comment by BIGWangYuDong** [comment]
LGTM now

**Review by hhaAndroid — APPROVED** [review]
